framework/data/BaseDataProvider.php

```diff
@@ -44,6 +50,20 @@
     private $_keys;
     private $_models;
     private $_totalCount;
+    
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (!isset($this->id)) {
+            if (self::$counter > 0) {
```

Comment:
I'd appreciate that, I'll open a new issue tomorrow for the refactoring, we can gather  some examples of counters within the framework and decide what features are needed.